pkg/controllermanager/controller/shoot/shoot_conditions_control.go

```diff
@@ -0,0 +1,148 @@
+}
+
+func (c *Controller) filterSeedForShootConditions(obj, oldObj, _ client.Object, deleted bool) bool {
+	seed, ok := obj.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	if oldObj != nil {
+		oldSeed, ok := oldObj.(*gardencorev1beta1.Seed)
+		if !ok {
+			return false
+		}
+
+		if !apiequality.Semantic.DeepEqual(seed.Status.Conditions, oldSeed.Status.Conditions) {
+			logger.Logger.Debugf("Seed %s conditions changed", seed.Name)
+			return true
+		}
+	}
+
+	if deleted {
+		logger.Logger.Debugf("Seed %s deleted", seed.Name)
+		return true
+	}
+
+	return false
+}
+
+// NewShootConditionsReconciler creates a reconcile.Reconciler that updates the conditions of a shoot that is registered as seed.
+func NewShootConditionsReconciler(logger logrus.FieldLogger, gardenClient client.Client, cfg *config.ShootConditionsControllerConfiguration) reconcile.Reconciler {
+	return &shootConditionsReconciler{
+		logger:       logger,
+		gardenClient: gardenClient,
+		cfg:          cfg,
+	}
+}
+
+type shootConditionsReconciler struct {
+	logger       logrus.FieldLogger
+	gardenClient client.Client
+	cfg          *config.ShootConditionsControllerConfiguration
+}
+
+func (r *shootConditionsReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.gardenClient.Get(ctx, request.NamespacedName, shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Infof("Object %q is gone, stop reconciling: %v", request.Name, err)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	// Get the seed this shoot is registered as
+	seed, err := r.getShootSeed(ctx, shoot)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Build new shoot conditions
+	// First remove all existing seed conditions and then then add the current seed conditions
+	// if the shoot is still registered as seed
+	seedConditionTypes := []gardencorev1beta1.ConditionType{
+		gardencorev1beta1.SeedBackupBucketsReady,
+		gardencorev1beta1.SeedBootstrapped,
+		gardencorev1beta1.SeedExtensionsReady,
+		gardencorev1beta1.SeedGardenletReady,
+	}
+	conditions := gardencorev1beta1helper.RemoveConditions(shoot.Status.Conditions, seedConditionTypes...)
+	if seed != nil {
+		conditions = gardencorev1beta1helper.MergeConditions(conditions, seed.Status.Conditions...)
+	}
+
+	// Update the shoot conditions if needed
+	if gardencorev1beta1helper.ConditionsNeedUpdate(shoot.Status.Conditions, conditions) {
+		r.logger.Debugf("Updating shoot %s conditions", kutil.ObjectName(shoot))
+		if err := r.updateShootConditions(ctx, shoot, conditions); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// If the shoot is still registered as seed, requeue after the configured sync period
+	if seed != nil {
+		return reconcile.Result{RequeueAfter: r.cfg.SyncPeriod.Duration}, nil
```

Comment:
If I see it correctly, the controller is already implemented very cleanly and doesn't need a period sync, as it watches the seeds for relevant condition changes.
I would drop the period sync interval and rather make sure we correctly handle cache resync events (i.e. update events with `old.resourceVersion==new.resourceVersion`).

Reply:
I tend to disagree. If a relevant seed event (the change in the seed conditions) is missed, the seed conditions in the shoot status will not be updated until the next cache resync (in 10 hours I believe?). I think this is too much. Even caching seeds in the filter function as proposed in https://github.com/gardener/remedy-controller/pull/43 will not help because seed events (unlike e.g. node events) are rather rare (I checked some landscape seeds that haven't been updated for a few hours, and then only updated because the health of an extension changed). So if we do need these conditions, we should make sure they are not outdated by more than e.g. 30 seconds or a minute, and the only reliable  way to achieve this IMO is to use a sync period.

Reply:
Also note that the sync period only kicks-in for shoots that are actually registered as seeds, otherwise it's never used.

Reply:
I get what you're saying. However, if we actually miss seed update events, it means, that we already have some connection issues to the API server. Hence, the shoot cache might be outdated as well, which means a regular sync might also act on outdated information and cause outdated seed conditions. 
The only way to work around this, is to issue direct reads every 30s, but I think we don't have to argue that this is not scalable. Even in that case, we might have connection issues and simply be unable to update seed conditions for a bit of time.

I think, completely missing relevant update events should be rare: if the reflector looses the watch connection, it will try to reestablish it and request watch events from the last observed resource version from the API server. Even if we might not get all the events that happened in between, we will get at least one update event per object, if something happened in between. This is pretty much eventually consistent.
Completely missing all relevant update events will only happen if there are some longer connection cutoffs to the API server. But as said, even then regular syncs don't help to make the conditions "fresher".

I think, the majority of k8s controllers doesn't have a regular sync interval and that's because it's generally not needed. They are all relying only on watch events and regular cache syncs.

Reply:
OK, thanks for explaining. I will remove the sync period then as suggested.

I have a couple of questions about this: 

> make sure we correctly handle cache resync events (i.e. update events with `old.resourceVersion==new.resourceVersion`).

* The resource version is never checked by the new reconciler, does this mean I don't need to do anything special to handle resync events?
* Where is the cache resync period actually configured (since g/g controllers are not yet fully based on `controller-runtime`, where the default value is 10h), and how can I configure it to a smaller value in order to test that they are handled?

Reply:
OK, found it: https://github.com/gardener/gardener/blob/7ff5a6ca8e385b0bfd0075597159f8799e73700b/pkg/utils/kubernetes/eventhandler.go#L144-L147
I wonder why these lines were added (I wrote this code but don't remember anymore) and could I simply remove them ... Any advice?

Reply:
Well, it still doesn't work even with the above lines removed. I added some more logs and it seems these cache resync events are never received, at least not for Seeds. Any help would be really appreciated.

Reply:
I think it's safe to remove. The condition will only fire for the periodic cache syncs and that's exactly what we want.

Reply:
OK, I will remove it. `ControlledResourceEventHandler` is only used here and in the  `ManagedSeed` and `ManagedSeedSet` controllers. The latter is not really used in production, and I don't see how the former could be adversely affected by such a change.

Reply:
I removed the sync period and removed the above lines from `ControlledResourceEventHandler`.